client/executor/wasmtime/src/instance_wrapper.rs

```diff
@@ -0,0 +1,245 @@
+		.clone();
+
+	Ok(memory)
+}
+
+/// Functions realted to memory.
+impl InstanceWrapper {
+	/// Read data from a slice of memory into a destination buffer.
+	///
+	/// Returns an error if the read would go out of the memory bounds.
+	pub fn read_memory_into(&self, address: Pointer<u8>, dest: &mut [u8]) -> Result<()> {
+		unsafe {
+			// This should be safe since we don't grow up memory while caching this reference and
+			// we give up the reference before returning from this function.
+			let memory = self.memory_as_slice();
+
+			let range = util::checked_range(address.into(), dest.len(), memory.len())
+				.ok_or_else(|| Error::Other("memory read is out of bounds".into()))?;
+			dest.copy_from_slice(&memory[range]);
+			Ok(())
+		}
+	}
+
+	/// Write data to a slice of memory.
+	///
+	/// Returns an error if the write would go out of the memory bounds.
+	pub fn write_memory_from(&self, address: Pointer<u8>, data: &[u8]) -> Result<()> {
+		unsafe {
+			// This should be safe since we don't grow up memory while caching this reference and
+			// we give up the reference before returning from this function.
+			let memory = self.memory_as_slice_mut();
+
+			let range = util::checked_range(address.into(), data.len(), memory.len())
+				.ok_or_else(|| Error::Other("memory write is out of bounds".into()))?;
+			&mut memory[range].copy_from_slice(data);
+			Ok(())
+		}
+	}
+
+	/// Allocate some memory of the given size. Returns pointer to the allocated memory region.
+	///
+	/// Returns `Err` in case memory cannot be allocated. Refer to the allocator documentation
+	/// to get more details.
+	pub fn allocate(
+		&self,
+		allocator: &mut sc_executor_common::allocator::FreeingBumpHeapAllocator,
+		size: WordSize,
+	) -> Result<Pointer<u8>> {
+		unsafe {
+			// This should be safe since we don't grow up memory while caching this reference and
+			// we give up the reference before returning from this function.
+			let memory = self.memory_as_slice_mut();
+
+			allocator.allocate(memory, size)
+		}
+	}
+
+	/// Deallocate the memory pointed by the given pointer.
+	///
+	/// Returns `Err` in case the given memory region cannot be deallocated.
+	pub fn deallocate(
+		&self,
+		allocator: &mut sc_executor_common::allocator::FreeingBumpHeapAllocator,
+		ptr: Pointer<u8>,
+	) -> Result<()> {
+		unsafe {
+			// This should be safe since we don't grow up memory while caching this reference and
+			// we give up the reference before returning from this function.
+			let memory = self.memory_as_slice_mut();
+
+			allocator.deallocate(memory, ptr)
+		}
+	}
+
+	/// Returns linear memory of the wasm instance as a slice.
+	///
+	/// # Safety
+	///
+	/// Wasmtime doesn't provide comprehensive documentation about the exact behavior of the data
+	/// pointer. If a dynamic style heap is used the base pointer of the heap can change. Since
+	/// growing, we cannot guarantee the lifetime of the returned slice reference.
```

Comment:
A better approach would be for this to be a function taking a callback, such that the reference cannot escape the callback.